File: otto-mobile/otto-mobile/MapMarkerSharePayload.swift
import SwiftUI

struct MapMarkerSharePayload: Equatable {
    enum PreviewKind: Equatable {
        case savedPlace
        case raceTrack
    }

    let title: String
    let subtitle: String?
    let latitude: Double?
    let longitude: Double?
    let savedPlaceId: String?
    let externalShareText: String
    let previewKind: PreviewKind

    static func savedPlace(
        id: String? = nil,
        name: String,
        addressSummary: String?,
        latitude: Double,
        longitude: Double
    ) -> MapMarkerSharePayload {
        MapMarkerSharePayload(
            title: name,
            subtitle: addressSummary,
            latitude: latitude,
            longitude: longitude,
            savedPlaceId: id,
            externalShareText: shareText(title: name, subtitle: addressSummary, lat: latitude, lng: longitude),
            previewKind: .savedPlace
        )
    }

    static func adhocPlace(
        name: String?,
        addressSummary: String?,
        latitude: Double,
        longitude: Double
    ) -> MapMarkerSharePayload {
        let title = (name?.trimmingCharacters(in: .whitespacesAndNewlines)).flatMap { $0.isEmpty ? nil : $0 }
            ?? (addressSummary?.trimmingCharacters(in: .whitespacesAndNewlines)).flatMap { $0.isEmpty ? nil : $0 }
            ?? String(localized: "chat_place_attachment_fallback_title")
        return MapMarkerSharePayload(
            title: title,
            subtitle: addressSummary,
            latitude: latitude,
            longitude: longitude,
            savedPlaceId: nil,
            externalShareText: shareText(title: title, subtitle: addressSummary, lat: latitude, lng: longitude),
            previewKind: .savedPlace
        )
    }

    static func raceTrack(
        name: String,
        locationLine: String?,
        latitude: Double?,
        longitude: Double?
    ) -> MapMarkerSharePayload {
        MapMarkerSharePayload(
            title: name,
            subtitle: locationLine,
            latitude: latitude,
            longitude: longitude,
            savedPlaceId: nil,
            externalShareText: shareText(title: name, subtitle: locationLine, lat: latitude, lng: longitude),
            previewKind: .raceTrack
        )
    }

    private static func shareText(title: String, subtitle: String?, lat: Double?, lng: Double?) -> String {
        var lines = [title]
        if let subtitle, !subtitle.isEmpty { lines.append(subtitle) }
        if let lat, let lng {
            lines.append(String(format: "%.5f, %.5f", lat, lng))
        }
        return lines.joined(separator: "\n")
    }
}

File: otto-mobile/otto-android-LEGACY-DO-NOT-OPEN/app/src/main/java/to/ottomot/driftd/MapMarkerSharePayload.kt
package to.ottomot.driftd

enum class MapMarkerSharePreviewKind {
    SavedPlace,
    RaceTrack,
}

data class MapMarkerSharePayload(
    val title: String,
    val subtitle: String?,
    val latitude: Double?,
    val longitude: Double?,
    val savedPlaceId: String? = null,
    val externalShareText: String,
    val previewKind: MapMarkerSharePreviewKind,
)

internal fun mapMarkerSharePayloadForSavedPlace(
    id: String? = null,
    name: String,
    addressSummary: String?,
    latitude: Double,
    longitude: Double,
): MapMarkerSharePayload =
    MapMarkerSharePayload(
        title = name,
        subtitle = addressSummary,
        latitude = latitude,
        longitude = longitude,
        savedPlaceId = id?.trim()?.takeIf { it.isNotEmpty() },
        externalShareText = mapMarkerShareText(name, addressSummary, latitude, longitude),
        previewKind = MapMarkerSharePreviewKind.SavedPlace,
    )

internal fun mapMarkerSharePayloadForAdhocPlace(
    name: String?,
    addressSummary: String?,
    latitude: Double,
    longitude: Double,
): MapMarkerSharePayload {
    val title =
        name?.trim()?.takeIf { it.isNotEmpty() }
            ?: addressSummary?.trim()?.takeIf { it.isNotEmpty() }
            ?: "Shared place"
    return MapMarkerSharePayload(
        title = title,
        subtitle = addressSummary,
        latitude = latitude,
        longitude = longitude,
        savedPlaceId = null,
        externalShareText = mapMarkerShareText(title, addressSummary, latitude, longitude),
        previewKind = MapMarkerSharePreviewKind.SavedPlace,
    )
}

internal fun mapMarkerSharePayloadForRaceTrack(
    name: String,
    locationLine: String?,
    latitude: Double?,
    longitude: Double?,
): MapMarkerSharePayload =
    MapMarkerSharePayload(
        title = name,
        subtitle = locationLine,
        latitude = latitude,
        longitude = longitude,
        savedPlaceId = null,
        externalShareText = mapMarkerShareText(name, locationLine, latitude, longitude),
        previewKind = MapMarkerSharePreviewKind.RaceTrack,
    )

internal fun mapMarkerShareText(
    title: String,
    subtitle: String?,
    lat: Double?,
    lng: Double?,
): String {
    val lines = mutableListOf(title)
    subtitle?.trim()?.takeIf { it.isNotEmpty() }?.let { lines.add(it) }
    if (lat != null && lng != null) {
        lines.add(String.format(java.util.Locale.US, "%.5f, %.5f", lat, lng))
    }
    return lines.joinToString("\n")
}
